molgenis-navigator/src/main/java/org/molgenis/navigator/copy/service/PackageCopier.java

```diff
@@ -0,0 +1,87 @@
+package org.molgenis.navigator.copy.service;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Streams.stream;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toSet;
+import static org.molgenis.data.meta.model.PackageMetadata.PACKAGE;
+import static org.molgenis.navigator.copy.service.LabelGenerator.generateUniqueLabel;
+
+import java.util.List;
+import java.util.Set;
+import org.molgenis.data.DataService;
+import org.molgenis.data.UnknownPackageException;
+import org.molgenis.data.meta.model.Package;
+import org.molgenis.data.meta.model.PackageMetadata;
+import org.molgenis.data.populate.IdGenerator;
+import org.molgenis.data.util.PackageUtils;
+import org.molgenis.navigator.copy.exception.RecursiveCopyException;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PackageCopier {
+
+  private final DataService dataService;
+  private final IdGenerator idGenerator;
+
+  PackageCopier(DataService dataService, IdGenerator idGenerator) {
+    this.dataService = requireNonNull(dataService);
+    this.idGenerator = requireNonNull(idGenerator);
+  }
+
+  public void copy(List<Package> packages, CopyState state) {
+    packages.forEach(pack -> copyPackage(pack, state));
+  }
+
+  private void copyPackage(Package pack, CopyState state) {
+    validateNotContainsItself(pack, state.targetPackage());
+    assignUniqueLabel(pack, state.targetPackage());
+    copyPackageRecursive(pack, state.targetPackage(), state);
+  }
+
+  private void copyPackageRecursive(Package pack, Package parent, CopyState state) {
+    state.entityTypesInPackages().addAll(newArrayList(pack.getEntityTypes()));
```

Comment:
new ArrayList<>();

Reply:
That's not possible here because I'm using the overloaded factory method `newArrayList(Iterable)`